crypto/bio/bio.c

```diff
@@ -186,6 +186,26 @@ int BIO_read(BIO *bio, void *buf, int len) {
   return ret;
 }
 
+int BIO_read_ex(BIO *bio, void *data, size_t data_len, size_t *read_bytes) {
+  if (bio == NULL || read_bytes == NULL) {
+    OPENSSL_PUT_ERROR(BIO, BIO_R_NULL_PARAMETER);
+    return 0;
+  }
+
+  int read_len = (int)data_len;
+  if (data_len > INT_MAX) {
+    read_len = INT_MAX;
+  }
+
+  int ret = BIO_read(bio, data, read_len);
+  *read_bytes = ret;
```

Comment:
Should this check to make sure that `BIO_read` didn't return a negative number first prior to setting `read_bytes` to that value? Otherwise you are potentially assigning a negative number in this case to an unsigned type. Then it looks like a weird scenario of "read X bytes but also had an error".

Maybe only update this value to `ret` in this `> 0` branch, otherwise set the value to `0` in the failure case.

Reply:
ahh good catch, I'll make the change.